core/src/main/java/org/infinispan/container/impl/AbstractInternalDataContainer.java

```diff
@@ -498,84 +498,157 @@ public void delete(K key, InternalCacheEntry<K, V> value, RemovalCause cause) {
+                           }
+                        });
+                     }
+                     return skipPassivation(orderer, key, future, operation);
                   }
+                  //falls through
                default:
+                  CompletionStage<Void> passivatedStage = passivator.passivateAsync(entry);
                   // This is a concurrent regular read - in which case we passivate just as normal
-                  return handleNotificationAndOrderer(key, value, passivator.passivateAsync(value), orderer, evictionManager, future);
+                  return handleNotificationAndOrderer(key, entry, passivatedStage, orderer, evictionManager, future);
             }
          });
       }
-      return handleNotificationAndOrderer(key, value, passivator.passivateAsync(value), orderer, evictionManager, future);
+      return handleNotificationAndOrderer(key, entry, passivator.passivateAsync(entry), orderer, evictionManager, future);
    }
 
-   private static CompletionStage<Void> skipEviction(NonBlockingOrderer orderer, Object key,
-         CompletableFuture<OPERATION> future, OPERATION op) {
+   private static CompletionStage<Void> skipPassivation(DataOperationOrderer orderer, Object key,
+         CompletableFuture<Operation> future, Operation op) {
       if (trace) {
          log.tracef("Skipping passivation for key %s due to %s", key, op);
       }
-      orderer.completeOperation(key, future, OPERATION.READ);
+      orderer.completeOperation(key, future, Operation.READ);
       return CompletableFutures.completedNull();
    }
 
    private static <K, V> CompletionStage<Void> handleNotificationAndOrderer(K key, InternalCacheEntry<K, V> value,
-         CompletionStage<Void> stage, NonBlockingOrderer orderer, EvictionManager<K, V> evictionManager,
-         CompletableFuture<OPERATION> future) {
+         CompletionStage<Void> stage, DataOperationOrderer orderer, EvictionManager<K, V> evictionManager,
+         CompletableFuture<Operation> future) {
       if (evictionManager != null) {
          stage = stage.thenCompose(ignore -> evictionManager.onEntryEviction(Collections.singletonMap(key, value)));
       }
       if (orderer != null) {
-         return stage.whenComplete((ignore, ignoreT) -> orderer.completeOperation(key, future, OPERATION.READ));
+         return stage.whenComplete((ignore, ignoreT) -> orderer.completeOperation(key, future, Operation.READ));
       }
       return stage;
    }
 
-   final class DefaultEvictionListener {
+   final class DefaultEvictionListener implements RemovalListener<K, InternalCacheEntry<K, V>> {
+      Map<Object, CompletableFuture<Void>> ensureEvictionDone = new ConcurrentHashMap<>();
 
       void onEntryChosenForEviction(K key, InternalCacheEntry<K, V> value) {
-         handleEviction(key, value, orderer, passivator.running(), evictionManager, AbstractInternalDataContainer.this);
+         CompletableFuture<Void> future = new CompletableFuture<>();
```

Comment:
Please add a comment here as well, e.g.

// Schedule an eviction to happen after the key lock is released

Reply:
Okay.